tests/integration/idmap.bats

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env bats
+
+load helpers
+
+function setup() {
+	requires root
+	requires_kernel 5.12
+	requires_idmap_fs /tmp
+
+	setup_debian
+
+	# Prepare source folders for bind mount
+	mkdir -p source-{1,2}/
+	touch source-{1,2}/foo.txt
+
+	# Use other owner for source-2
+	chown 1:1 source-2/foo.txt
+
+	mkdir -p rootfs/{proc,sys,tmp}
+	mkdir -p rootfs/tmp/mount-{1,2}
+	mkdir -p rootfs/mnt/bind-mount-{1,2}
+
+	update_config ' .linux.namespaces += [{"type": "user"}]
+		| .linux.uidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]
+		| .linux.gidMappings += [{"hostID": 100000, "containerID": 0, "size": 65536}]
+		| .process.args += ["-c", "stat -c =%u=%g= /tmp/mount-1/foo.txt"]
+		| .mounts += [
+					{
+						"source": "source-1/",
+						"destination": "/tmp/mount-1",
+						"options": ["bind"],
+						"uidMappings": [ {
+						                  "containerID": 0,
+						                  "hostID": 100000,
+						                  "size": 65536
+						                }
+						],
+						"gidMappings": [        {
+						                  "containerID": 0,
+						                  "hostID": 100000,
+						                  "size": 65536
+						                }
+						]
+					}
+				] '
+}
+
+function teardown() {
+	teardown_bundle
+}
+
+@test "simple idmap mount" {
+	runc run test_debian
+	[ "$status" -eq 0 ]
+	[[ "$output" == *"=0=0="* ]]
+}
+
+@test "idmap mount with propagation flag" {
+	update_config ' .process.args = ["sh", "-c", "findmnt -o PROPAGATION /tmp/mount-1"]'
+	# Add the shared option to the idmap mount
+	update_config ' .mounts |= map((select(.source == "source-1/") | .options += ["shared"]) // .)'
+
+	runc run test_debian
+	[ "$status" -eq 0 ]
+	[[ "$output" == *"shared"* ]]
+}
+
+@test "idmap mount with bind mount" {
+	update_config '   .mounts += [
+					{
+						"source": "source-2/",
+						"destination": "/tmp/mount-2",
+						"options": ["bind"]
+					}
+				] '
+
+	runc run test_debian
+	[ "$status" -eq 0 ]
+	[[ "$output" == *"=0=0="* ]]
+}
+
+@test "two idmap mounts with two bind mounts" {
+	update_config '   .process.args = ["sh", "-c", "stat -c =%u=%g= /tmp/mount-1/foo.txt /tmp/mount-2/foo.txt"]
```

Comment:
How about add some test cases about checking the write permission in a container? For example, touch an existing file, create a new file in a container and check the owner in the host.

Reply:
@lifubang Thanks! I've added a test for that, see the new "write to an idmap mount" test for more details.

I was on the fence if this was worth it or not, as it smells a little bit like testing the kernel feature rather than the runc feature, but the test is so simple and fast that I think there is no harm having it either.

Reply:
@lifubang oh, I just realize you said "and check the owner in the host". IMHO, that seems more testing the kernel feature than the runc feature.

I added a test that we write a file and check its ownership from inside the container. As this is an idmap mount, the only way it is shown correctly is because the UID/GID are the right on the host too. IMHO, that is enough to check what we are doing on runc.

The kernel side already has lot of tests for idmap mounts in xfstests (there are several suites here: https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git/tree/src/vfs, and there are some more tests in bash on that repo too).

But please feel free to re-open this if you think your concern is not addressed :)